### ISSUE TITLE
Enhanced Rate Limiting with a retryAfter option and IP Blocking Features for Improved Flexibility

### ISSUE THREAD

**Issue by hgouveia** [express-rate-limit/express-rate-limit#441]
### Description

Hello,

The main goal is to be able to set a different 'retry after' time. Additionally, I'd like the possibility to permanently ban certain IPs.

I propose having separate configuration parameters, something like this:

```js
app.use(rateLimit({
   windowMs: 1000, // 1seg
   limit: 10, // 10 req per sec
   /// NEW
   retryAfter:  60 * 60 * 1000  or async (req, ip) => number,   // retry in 1h
   notAllowed: [ '123.123.1.1']  or async  (req, ip) => bool
}):
```

Having the possibility to accept a function would make the system more flexible, especially in cases where the value varies depending on the user or IP.

### Why

There are many cases where you might want to allow a certain number of requests within a specific time window. However, this doesn't necessarily mean that access should be regained immediately after that same time window.

The main problem is that if I set `windowMs=1000` (1 second), users will be able to retry after just 1 second. But what I'd like to do is restrict this to a different time frame, such as 1 hour. Unfortunately, this isn't currently possible.

Additionally, I would like an option to block certain IPs. While I could create a separate middleware to check IPs and block them, it would be more suitable to leverage this library's existing advanced checks, such as proxy detection.

### Alternatives

N/A

**Comment by nfriedly** [comment]
Hi @hgouveia,

> The main goal is to be able to set a different 'retry after' time. 

That's a good idea.

You could do the extended Retry-After part easily enough by setting `headers: false` and then having a custom `handler()` function that sets the headers, but it would advisory only, the library wouldn't enforce it.

Enforcing it is going to be a bit harder though, due to the current Store interface design. Let me think about that a bit more, though.

Another option is to create a couple of instances of express-rate-limit. The first one could allow 10 requests per second, and the second one could be 1000 per day or something like that. I know that's not quite what you asked for, but it could address some of the need.

> Additionally, I'd like the possibility to permanently ban certain IPs.

I'm somewhat against the idea of permanently blocking IPs because most people are on dynamic IPs that change over time, so the blocks end up applying to people unrelated to the initial abuse. 

Additionally proxies have legitimate uses, so I feel like blocking them outright throws out the good with the bad, and it is also getting farther away from the intended purpose of this library. 

(Also, I've personally had my IP mislabeled as a proxy for the [dumbest of reasons](https://github.com/NginxProxyManager/nginx-proxy-manager/discussions/2127#discussioncomment-6326517), and it took months to resolve. During that time, all kinds of things just didn't work, including my wifi garage door opener and the USPS website.)

**Comment by gamemaker1** [comment]
Sorry for the late reply, but you can block certain IPs if you really need to, using the `skip` option. You can see an example of it being used as a blocklist [here](https://express-rate-limit.mintlify.app/reference/configuration#skip).

**Comment by Zaokret** [comment]
> Sorry for the late reply, but you can block certain IPs if you really need to, using the `skip` option. You can see an example of it being used as a blocklist [here](https://express-rate-limit.mintlify.app/reference/configuration#skip).

`skip` will disable limiter in the specified case that evaluates to true

**Comment by gamemaker1** [comment]
Ah you're right, what I suggested in my earlier comment was wrong. Sorry about that.

You can use the `skip` option as an allowlist (to not apply the rate limit for certain clients) and the [`limit`](https://express-rate-limit.mintlify.app/reference/configuration#limit) option as a blocklist (to outright block all requests from certain clients) in v7+, like so:

```ts
const allowlist = ['192.168.0.56', '192.168.0.21']
const blocklist = ['192.168.0.45']

const limiter = rateLimit({
	// ...
	skip: (req, res) => allowlist.includes(req.ip),
	limit: (req, res) => blocklist.includes(req.ip) ? 0 : 5
})
```

This is possible, since as of version 7.0.0, setting max to zero will no longer disable the rate limiter - instead, it will ‘block’ all requests to that endpoint.

**Comment by nfriedly** [comment]
It sounds like you have a good solution, so I'm going to go ahead and close this.

**Comment by hgouveia** [comment]
> It sounds like you have a good solution, so I'm going to go ahead and close this.

Hello @nfriedly , what @gamemaker1  suggest just solves the blocking feature, but no the main feature about be able to set different 'retry after' time in my original post

could we at least create a copy of this issue with only that part?

**Comment by nfriedly** [comment]
Oh, sorry, I should have read that more closely. I'll re-open it.

**Comment by aarifkhan7** [comment]
Hello @nfriedly,

I'm a beginner interested in contributing to open source, and this issue seems good for me. Could you please assign it to me?

Thanks!

**Comment by nfriedly** [comment]
Hi @aarifkhan7, thanks! I went ahead and assigned it to you.

For future reference, we generally don't use the "Assign" feature in this repo - a comment mentioning that you're working on it, perhaps with an outline of what you have in mind, would be sufficient.

**Comment by aarifkhan7** [comment]
Thank you for the information, @nfriedly. I will take this into account.

I see two possible approaches to implement the Retry-After option feature. One is to allow the user to define a retryAfter option (of type `ValueDeterminingMiddleware`), which will return the number of seconds to wait. The other approach is to accept a retryAfter option (boolean), if set, it calculates the wait time from the `RateLimitInfo.resetTime` object. The wait time can then be included in the request headers.

Please let me know your thoughts on this approach.

**Comment by gamemaker1** [comment]
I think that if we do implement this feature, we should go with the first option; it fits in well with the current API we have.

However, I'm not sure whether I understand the need for this feature correctly:

> There are many cases where you might want to allow a certain number of requests within a specific time window. However, this doesn't necessarily mean that access should be regained immediately after that same time window.
> 
> The main problem is that if I set windowMs=1000 (1 second), users will be able to retry after just 1 second. But what I'd like to do is restrict this to a different time frame, such as 1 hour. Unfortunately, this isn't currently possible.

Even with a different `Retry-After` time set in the headers, the users can choose to disregard it and try again after just one second. If you wish to restrict retrying to a different time frame, doesn't it make more sense to set the window to one hour?

**Comment by hgouveia** [comment]
> Even with a different `Retry-After` time set in the headers, the users can choose to disregard it and try again after just one second. If you wish to restrict retrying to a different time frame, doesn't it make more sense to set the window to one hour?

They are two different things for me. Let me explain again.

We use `windowMs` to allow a certain number of requests within that window. For example, `{windowMs: 1000, limit: 10}` means you can make 10 requests in 1 second. However, this does not mean the client should retry after 1 second as well (current behavior). I would like to decide that if a client hits the rate limit, they need to retry after 1 hour. This can be for whatever reason I decide, probably because the client reached the rate limit too many times, etc. If I set `windowMs` to 1 hour, it is not the same. I am saying the client can make 10 requests in 1 hour.

I hope this makes it clearer.

**Comment by gamemaker1** [comment]
That makes it much clearer, thank you.

The use case you describe is not possible with the way the stores work at the moment, and the `Retry-After` feature will not be the solution, since the header is not enforceable. However, I think we can still go ahead with implementing the `Retry-After` feature anyways, since it seems independently useful.

@aarifkhan7 would you like to make a PR using the first option you suggested (via `ValueDeterminingMiddleware`)?

**Comment by nfriedly** [comment]
Lets choose a different name for this feature, maybe something like "cool-down period" to describe the time between a user hitting the rate limit and when they are allowed additional requests. Right now this is always equal to the time remaining in the current window, but we could conceivably change things to make this be configurable to a separate, larger value.

The `Retry-After` header will correctly describe this cool-down period, but it will also be enforced by the server, so it's not just the `Retry-After` header.

The most difficult challenge with the cool-down period is store support. Most (all?) of the existing stores have built-in optimizations to take advantage of the fact that the reset time is currently the same regardless of whether the max was hit or not. To add this feature, we'll need some way of not resetting certain users when they normally would be. That's the part I'm not sure how to accomplish without breaking a lot of existing things.

Also

> This can be for whatever reason I decide, probably because the client reached the rate limit too many times, etc. 

That adds a bit of additional complexity, as now we need to also track how many times the user has reached the limit.

These needs are getting different enough from the design of this library that you might be better off using a different one that was designed to be a bit more flexible. Perhaps https://www.npmjs.com/package/rate-limiter-flexible or https://www.npmjs.com/package/express-brute could meet your needs?

**Comment by aarifkhan7** [comment]
I agree with you, @nfriedly. As @gamemaker1 mentioned and I also believe, including the `Retry-After` header would add semantic value. I'll work on creating the PR this weekend.

**Comment by bit-64kce** [comment]
Please could i try to work on this feat as issue has not been closed ?

**Comment by nfriedly** [comment]
Hi @NwankwoKc, what changes do you have in mind?
